## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [1]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [2]:
# Write your code for importing the csv file here
movies = pd.read_csv("E:/GDrive_MaheBil_Sync/Viki/IIIT_B/Preparatory_Course/Python/Movies.csv")
movies.head(10)

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   
5  Color          Sam Raimi                   392.0     156.0   
6  Color       Nathan Greno                   324.0     100.0   
7  Color        Joss Whedon                   635.0     141.0   
8  Color        David Yates                   375.0     153.0   
9  Color        Zack Snyder                   673.0     183.0   

   director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                      0.0                   855.0   Joel David Moore   
1                    563.0                  1000.0      Orlando Bloom   
2                      0.0                   161.0       Rory Kinnear   
3                  22000.0                 23000.0     Christian Bale   
4                    475.0                   530.0    Samantha Morton   
5                      0.0                  4000.0       James Franco   
6                     15.0                   284.0       Donna Murphy   
7                      0.0                 19000.0  Robert Downey Jr.   
8                    282.0                 10000.0   Daniel Radcliffe   
9                      0.0                  2000.0       Lauren Cohan   

   actor_1_facebook_likes        gross  \
0                  1000.0  760505847.0   
1                 40000.0  309404152.0   
2                 11000.0  200074175.0   
3                 27000.0  448130642.0   
4                   640.0   73058679.0   
5                 24000.0  336530303.0   
6                   799.0  200807262.0   
7                 26000.0  458991599.0   
8                 25000.0  301956980.0   
9                 15000.0  330249062.0   

                                              genres  ...  \
0                    Action|Adventure|Fantasy|Sci-Fi  ...   
1                           Action|Adventure|Fantasy  ...   
2                          Action|Adventure|Thriller  ...   
3                                    Action|Thriller  ...   
4                            Action|Adventure|Sci-Fi  ...   
5                           Action|Adventure|Romance  ...   
6  Adventure|Animation|Comedy|Family|Fantasy|Musi...  ...   
7                            Action|Adventure|Sci-Fi  ...   
8                   Adventure|Family|Fantasy|Mystery  ...   
9                            Action|Adventure|Sci-Fi  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                738.0  English      USA           PG-13  263700000.0   
5               1902.0  English      USA           PG-13  258000000.0   
6                387.0  English      USA              PG  260000000.0   
7               1117.0  English      USA           PG-13  250000000.0   
8                973.0  English       UK              PG  250000000.0   
9               3018.0  English      USA           PG-13  250000000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4      2012.0                  632.0        6.6          2.35   
5      2007.0                11000.0        6.2          2.35   
6      2010.0                  553.0        7.8          1.85   
7      2015.0           

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [3]:

#Question1 
# Write your code for inspection here
movies.shape

(3853, 28)

In [4]:

#Question2
# Write your code for inspection here
print((movies.isnull().sum()>0).sum())

12


#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [5]:
#Question 3
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations
movies.drop(['color','director_facebook_likes','actor_1_facebook_likes','actor_2_facebook_likes','actor_3_facebook_likes','actor_2_name','cast_total_facebook_likes','actor_3_name','duration','facenumber_in_poster','content_rating','country','movie_imdb_link','aspect_ratio',
'plot_keywords'], axis=1,inplace = True)

#print(movies)



#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

In [6]:
movies.shape

(3853, 13)

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [7]:
round(100*(movies.isnull().sum()/len(movies.index)),2) 


director_name             0.00
num_critic_for_reviews    0.03
gross                     0.00
genres                    0.00
actor_1_name              0.00
movie_title               0.00
num_voted_users           0.00
num_user_for_reviews      0.00
language                  0.08
budget                    0.00
title_year                0.00
imdb_score                0.00
movie_facebook_likes      0.00
dtype: float64

#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

In [8]:
# Write you code here

a = round(100*(movies.isnull().sum()/len(movies.index)),2)
a.sort_values(ascending = False) 

language                  0.08
num_critic_for_reviews    0.03
movie_facebook_likes      0.00
imdb_score                0.00
title_year                0.00
budget                    0.00
num_user_for_reviews      0.00
num_voted_users           0.00
movie_title               0.00
actor_1_name              0.00
genres                    0.00
gross                     0.00
director_name             0.00
dtype: float64

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [9]:
# Write your code for filling the NaN values in the 'language' column here

movies['language'] = movies['language'].astype('str') #NaN is float type so convert it into String value.
movies['language'] = movies.language.apply(lambda x : "English" if x=='nan' else x)
movies
#print((movies[movies["director_name"] == 'Christopher Cain'] )["language"])


director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
1        Gore Verbinski                   302.0  309404152.0   
2            Sam Mendes                   602.0  200074175.0   
3     Christopher Nolan                   813.0  448130642.0   
4        Andrew Stanton                   462.0   73058679.0   
...                 ...                     ...          ...   
3848      Shane Carruth                   143.0     424760.0   
3849   Neill Dela Llana                    35.0      70071.0   
3850   Robert Rodriguez                    56.0    2040920.0   
3851       Edward Burns                    14.0       4584.0   
3852           Jon Gunn                    43.0      85222.0   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Christoph Waltz   
3                         Action|Thriller        Tom Hardy   
4                 Action|Adventure|Sci-Fi     Daryl Sabara   
...                                   ...              ...   
3848                Drama|Sci-Fi|Thriller    Shane Carruth   
3849                             Thriller      Ian Gamazon   
3850  Action|Crime|Drama|Romance|Thriller  Carlos Gallardo   
3851                         Comedy|Drama      Kerry Bishé   
3852                          Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
4                                  John Carter            212204   
...                                         ...              ...   
3848                                    Primer             72639   
3849                                    Cavite               589   
3850                               El Mariachi             52055   
3851                                 Newlyweds              1338   
3852                         My Date with Drew              4285   

      num_user_for_reviews language       budget  title_year  imdb_score  \
0                   3054.0  English  237000000.0      2009.0         7.9   
1                   1238.0  English  300000000.0      2007.0         7.1   
2                    994.0  English  245000000.0      2015.0         6.8   
3                   2701.0  English  250000000.0      2012.0         8.5   
4                    738.0  English  263700000.0      2012.0         6.6   
...                    ...      ...          ...         ...         ...   
3848                 371.0  English       7000.0      2004.0         7.0   
3849                  35.0  English       7000.0      2005.0         6.3   
3850                 130.0  Spanish       7000.0      1992.0         6.9   
3851                  14.0  English       9000.0      2011.0         6.4   
3852                  84.0  English       1100.0      2004.0         6.6   

      movie_facebook_likes  
0                    33000  
1                        0  
2                    85000  
3                   164000  
4                    24000  
...                    ...  
3848                 19000  
3849                    74  
3850                     0  
3851                   413  
3852                   456  

[3853 rows x 13 columns]

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

In [10]:
(movies['language'] == 'English').sum()

3674

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [11]:
# Write your code for unit conversion here
#print(movies["gross"]/10**6)

#print(movies['budget']/10**6)
movies["gross_million"] = movies["gross"]/10**6
movies["budget_million"] = movies["budget"]/10**6
movies


director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
1        Gore Verbinski                   302.0  309404152.0   
2            Sam Mendes                   602.0  200074175.0   
3     Christopher Nolan                   813.0  448130642.0   
4        Andrew Stanton                   462.0   73058679.0   
...                 ...                     ...          ...   
3848      Shane Carruth                   143.0     424760.0   
3849   Neill Dela Llana                    35.0      70071.0   
3850   Robert Rodriguez                    56.0    2040920.0   
3851       Edward Burns                    14.0       4584.0   
3852           Jon Gunn                    43.0      85222.0   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Christoph Waltz   
3                         Action|Thriller        Tom Hardy   
4                 Action|Adventure|Sci-Fi     Daryl Sabara   
...                                   ...              ...   
3848                Drama|Sci-Fi|Thriller    Shane Carruth   
3849                             Thriller      Ian Gamazon   
3850  Action|Crime|Drama|Romance|Thriller  Carlos Gallardo   
3851                         Comedy|Drama      Kerry Bishé   
3852                          Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
4                                  John Carter            212204   
...                                         ...              ...   
3848                                    Primer             72639   
3849                                    Cavite               589   
3850                               El Mariachi             52055   
3851                                 Newlyweds              1338   
3852                         My Date with Drew              4285   

      num_user_for_reviews language       budget  title_year  imdb_score  \
0                   3054.0  English  237000000.0      2009.0         7.9   
1                   1238.0  English  300000000.0      2007.0         7.1   
2                    994.0  English  245000000.0      2015.0         6.8   
3                   2701.0  English  250000000.0      2012.0         8.5   
4                    738.0  English  263700000.0      2012.0         6.6   
...                    ...      ...          ...         ...         ...   
3848                 371.0  English       7000.0      2004.0         7.0   
3849                  35.0  English       7000.0      2005.0         6.3   
3850                 130.0  Spanish       7000.0      1992.0         6.9   
3851                  14.0  English       9000.0      2011.0         6.4   
3852                  84.0  English       1100.0      2004.0         6.6   

      movie_facebook_likes  gross_million  budget_million  
0                    33000     760.505847        237.0000  
1                        0     309.404152        300.0000  
2                    85000     200.074175        245.0000  
3                   164000     448.130642        250.0000  
4                    24000      73.058679        263.7000  
...                    ...            ...             ...  
3848                 19000       0.424760          0.0070  
3849                    74       0.070071          0.0070  
3850                     0       2.040920          0.0070  
3851                   413       0.004584          0.0090  
3852                   456       0.085222          0.0011  

[3853 rows x 15 columns]

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [12]:
# Write your code for creating the profit column here
movies["Profit"] = movies["gross_million"]-movies["budget_million"]


In [13]:
# Write your code for sorting the dataframe here
top_10 = movies.sort_values("Profit", axis = 0, ascending = False)




In [14]:
top10 = top_10.head(10)# Write your code to get the top 10 profiting movies here
top10

director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
28      Colin Trevorrow                   644.0  652177271.0   
25        James Cameron                   315.0  658672302.0   
2704       George Lucas                   282.0  460935665.0   
2748   Steven Spielberg                   215.0  434949459.0   
16          Joss Whedon                   703.0  623279547.0   
482        Roger Allers                   186.0  422783777.0   
230        George Lucas                   320.0  474544677.0   
64    Christopher Nolan                   645.0  533316061.0   
419           Gary Ross                   673.0  407999255.0   

                                        genres         actor_1_name  \
0              Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
28            Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
25                               Drama|Romance    Leonardo DiCaprio   
2704           Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
2748                             Family|Sci-Fi         Henry Thomas   
16                     Action|Adventure|Sci-Fi      Chris Hemsworth   
482   Adventure|Animation|Drama|Family|Musical    Matthew Broderick   
230            Action|Adventure|Fantasy|Sci-Fi      Natalie Portman   
64                 Action|Crime|Drama|Thriller       Christian Bale   
419            Adventure|Drama|Sci-Fi|Thriller    Jennifer Lawrence   

                                     movie_title  num_voted_users  \
0                                        Avatar            886204   
28                               Jurassic World            418214   
25                                      Titanic            793059   
2704         Star Wars: Episode IV - A New Hope            911097   
2748                 E.T. the Extra-Terrestrial            281842   
16                                 The Avengers            995415   
482                               The Lion King            644348   
230   Star Wars: Episode I - The Phantom Menace            534658   
64                              The Dark Knight           1676169   
419                            The Hunger Games            701607   

      num_user_for_reviews language       budget  title_year  imdb_score  \
0                   3054.0  English  237000000.0      2009.0         7.9   
28                  1290.0  English  150000000.0      2015.0         7.0   
25                  2528.0  English  200000000.0      1997.0         7.7   
2704                1470.0  English   11000000.0      1977.0         8.7   
2748                 515.0  English   10500000.0      1982.0         7.9   
16                  1722.0  English  220000000.0      2012.0         8.1   
482                  656.0  English   45000000.0      1994.0         8.5   
230                 3597.0  English  115000000.0      1999.0         6.5   
64                  4667.0  English  185000000.0      2008.0         9.0   
419                 1959.0  English   78000000.0      2012.0         7.3   

      movie_facebook_likes  gross_million  budget_million      Profit  
0                    33000     760.505847           237.0  523.505847  
28                  150000     652.177271           150.0  502.177271  
25                   26000     658.672302           200.0  458.672302  
2704                 33000     460.935665            11.0  449.935665  
2748                 34000     434.949459            10.5  424.449459  
16                  123000     623.279547           220.0  403.279547  
482                  17000     422.783777            45.0  377.783777  
230                  13000     474.544677           115.0  359.544677  
64                   37000     533.316061           185.0  348.316061  
419                 140000     407.999255            78.0  329.999255

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

In [15]:

top10["Rank"] = top10["Profit"].rank(ascending = False) 
top10[top10["Rank"] == 5]



D:\Anaconda\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


director_name  num_critic_for_reviews        gross         genres  \
2748  Steven Spielberg                   215.0  434949459.0  Family|Sci-Fi   

      actor_1_name                  movie_title  num_voted_users  \
2748  Henry Thomas  E.T. the Extra-Terrestrial            281842   

      num_user_for_reviews language      budget  title_year  imdb_score  \
2748                 515.0  English  10500000.0      1982.0         7.9   

      movie_facebook_likes  gross_million  budget_million      Profit  Rank  
2748                 34000     434.949459            10.5  424.449459   5.0

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [16]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'
IMDb_Top_250 = movies.sort_values(by = 'imdb_score', ascending = False)
IMDb_Top_250 = IMDb_Top_250.loc[IMDb_Top_250.num_voted_users > 25000]
IMDb_Top_250 = IMDb_Top_250.iloc[:250, ]
IMDb_Top_250['Rank'] = range(1,251)
IMDb_Top_250


director_name  num_critic_for_reviews        gross  \
1795        Frank Darabont                   199.0   28341469.0   
3016  Francis Ford Coppola                   208.0  134821952.0   
2543  Francis Ford Coppola                   149.0   57300000.0   
64       Christopher Nolan                   645.0  533316061.0   
325          Peter Jackson                   328.0  377019252.0   
...                    ...                     ...          ...   
2708      David O. Russell                   410.0   93571803.0   
22           Peter Jackson                   509.0  258355354.0   
1612         James Mangold                   291.0  119518352.0   
3237          Duncan Jones                   415.0    5009677.0   
874         Andrew Adamson                   212.0  267652016.0   

                                         genres           actor_1_name  \
1795                                Crime|Drama         Morgan Freeman   
3016                                Crime|Drama              Al Pacino   
2543                                Crime|Drama         Robert De Niro   
64                  Action|Crime|Drama|Thriller         Christian Bale   
325              Action|Adventure|Drama|Fantasy          Orlando Bloom   
...                                         ...                    ...   
2708                      Biography|Drama|Sport         Christian Bale   
22                            Adventure|Fantasy           Aidan Turner   
1612              Biography|Drama|Music|Romance  Sandra Ellis Lafferty   
3237                       Drama|Mystery|Sci-Fi           Kevin Spacey   
874   Adventure|Animation|Comedy|Family|Fantasy       Kathleen Freeman   

                                         movie_title  num_voted_users  \
1795                       The Shawshank Redemption           1689764   
3016                                  The Godfather           1155770   
2543                         The Godfather: Part II            790926   
64                                  The Dark Knight           1676169   
325   The Lord of the Rings: The Return of the King           1215718   
...                                              ...              ...   
2708                                    The Fighter            275869   
22              The Hobbit: The Desolation of Smaug            483540   
1612                                  Walk the Line            188637   
3237                                           Moon            260607   
874                                           Shrek            467113   

      num_user_for_reviews language       budget  title_year  imdb_score  \
1795                4144.0  English   25000000.0      1994.0         9.3   
3016                2238.0  English    6000000.0      1972.0         9.2   
2543                 650.0  English   13000000.0      1974.0         9.0   
64                  4667.0  English  185000000.0      2008.0         9.0   
325                 3189.0  English   94000000.0      2003.0         8.9   
...                    ...      ...          ...         ...         ...   
2708                 389.0  English   25000000.0      2010.0         7.9   
22                   951.0  English  225000000.0      2013.0         7.9   
1612                 815.0  English   28000000.0      2005.0         7.9   
3237                 485.0  English    5000000.0      2009.0         7.9   
874                  945.0  English   60000000.0      2001.0         7.9   

      movie_facebook_likes  gross_million  budget_million      Profit  Rank  
1795                108000      28.341469            25.0    3.341469     1  
3016                 43000     134.821952             6.0  128.821952     2  
2543                 14000      57.300000            13.0   44.300000     3  
64                   37000     533.316061           185.0  348.316061     4  
325                  16000     377.019252            94.0  283.019252     5  
...                    ...            ...             ...    

#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

In [17]:

range1 =len(IMDb_Top_250[(IMDb_Top_250['imdb_score'] >= 7.5) & (IMDb_Top_250['imdb_score'] <=8)] )
print("range1",range1)
range2 =len(IMDb_Top_250[(IMDb_Top_250['imdb_score'] >= 8) & (IMDb_Top_250['imdb_score'] <=8.5)] )
print("range2",range2)
range3 =len(IMDb_Top_250[(IMDb_Top_250['imdb_score'] >= 8.5) & (IMDb_Top_250['imdb_score'] <=9)] )
print("range3",range3)
range4 =len(IMDb_Top_250[(IMDb_Top_250['imdb_score'] >= 9) & (IMDb_Top_250['imdb_score'] <=9.5)] )
print("range4",range4)
range5 =len(IMDb_Top_250[(IMDb_Top_250['imdb_score'] >= 9.5) & (IMDb_Top_250['imdb_score'] <=10)] )
print("range5",range5)

range1 98
range2 177
range3 44
range4 4
range5 0


**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [18]:
# Write your code for creating three new dataframes here


Meryl_Streep = movies[movies['actor_1_name'] =='Meryl Streep'] # Include all movies in which Meryl_Streep is the lead
Meryl_Streep

director_name  num_critic_for_reviews        gross  \
392     Nancy Meyers                   187.0  112703470.0   
1038   Curtis Hanson                    42.0   46815748.0   
1132     Nora Ephron                   252.0   94125426.0   
1322   David Frankel                   208.0  124732962.0   
1390  Robert Redford                   227.0   14998070.0   
1471  Sydney Pollack                    66.0   87100000.0   
1514   David Frankel                   234.0   63536011.0   
1563   Carl Franklin                    64.0   23209440.0   
1784  Stephen Daldry                   174.0   41597830.0   
2500  Phyllida Lloyd                   331.0   29959436.0   
2793   Robert Altman                   211.0   20338609.0   

                               genres  actor_1_name  \
392              Comedy|Drama|Romance  Meryl Streep   
1038  Action|Adventure|Crime|Thriller  Meryl Streep   
1132          Biography|Drama|Romance  Meryl Streep   
1322             Comedy|Drama|Romance  Meryl Streep   
1390               Drama|Thriller|War  Meryl Streep   
1471          Biography|Drama|Romance  Meryl Streep   
1514             Comedy|Drama|Romance  Meryl Streep   
1563                            Drama  Meryl Streep   
1784                    Drama|Romance  Meryl Streep   
2500          Biography|Drama|History  Meryl Streep   
2793               Comedy|Drama|Music  Meryl Streep   

                    movie_title  num_voted_users  num_user_for_reviews  \
392           It's Complicated             69860                 214.0   
1038            The River Wild             32544                  69.0   
1132             Julie & Julia             79264                 277.0   
1322     The Devil Wears Prada            286178                 631.0   
1390           Lions for Lambs             41170                 298.0   
1471             Out of Africa             52339                 200.0   
1514              Hope Springs             34258                 178.0   
1563            One True Thing              9283                 112.0   
1784                 The Hours            102123                 660.0   
2500             The Iron Lady             82327                 350.0   
2793  A Prairie Home Companion             19655                 280.0   

     language      budget  title_year  imdb_score  movie_facebook_likes  \
392   English  85000000.0      2009.0         6.6                     0   
1038  English  45000000.0      1994.0         6.3                     0   
1132  English  40000000.0      2009.0         7.0                 13000   
1322  English  35000000.0      2006.0         6.8                     0   
1390  English  35000000.0      2007.0         6.2                     0   
1471  English  31000000.0      1985.0         7.2                     0   
1514  English  30000000.0      2012.0         6.3                     0   
1563  English  30000000.0      1998.0         7.0                   592   
1784  English  25000000.0      2002.0         7.6                     0   
2500  English  13000000.0      2011.0         6.4                 18000   
2793  English  10000000.0      2006.0         6.8                   683   

      gross_million  budget_million     Profit  
392      112.703470            85.0  27.703470  
1038      46.815748            45.0   1.815748  
1132      94.125426            40.0  54.125426  
1322     124.732962            35.0  89.732962  
1390      14.998070            35.0 -20.001930  
1471      87.100000            31.0  56.100000  
1514      63.536011            30.0  33.536011  
1563      23.209440            30.0  -6.790560  
1784      41.597830            25.0  16.597830  
2500      29.959436            13.0  16.959436  
2793      20.338609            10.0  10.338609

In [19]:
Leo_Caprio = movies[movies['actor_1_name'] =='Leonardo DiCaprio']# Include all movies in which Leo_Caprio is the lead
Leo_Caprio

director_name  num_critic_for_reviews        gross  \
25            James Cameron                   315.0  658672302.0   
49             Baz Luhrmann                   490.0  144812796.0   
94        Christopher Nolan                   642.0  292568851.0   
173   Alejandro G. Iñárritu                   556.0  183635922.0   
246         Martin Scorsese                   267.0  102608827.0   
283       Quentin Tarantino                   765.0  162804648.0   
293            Edward Zwick                   166.0   57366262.0   
294         Martin Scorsese                   606.0  116866727.0   
312         Martin Scorsese                   233.0   77679638.0   
347         Martin Scorsese                   352.0  132373442.0   
431         Martin Scorsese                   490.0  127968405.0   
608            Ridley Scott                   238.0   39380442.0   
859        Steven Spielberg                   194.0  164435221.0   
934             Danny Boyle                   118.0   39778599.0   
1046             Sam Mendes                   323.0   22877808.0   
1336        Randall Wallace                    83.0   56876365.0   
1364         Clint Eastwood                   392.0   37304950.0   
1457              Sam Raimi                    63.0   18636537.0   
1909             Jerry Zaks                    45.0   12782508.0   
2484           Baz Luhrmann                   106.0   46338728.0   
3025           Baz Luhrmann                   490.0  144812796.0   

                                genres       actor_1_name  \
25                       Drama|Romance  Leonardo DiCaprio   
49                       Drama|Romance  Leonardo DiCaprio   
94    Action|Adventure|Sci-Fi|Thriller  Leonardo DiCaprio   
173   Adventure|Drama|Thriller|Western  Leonardo DiCaprio   
246                    Biography|Drama  Leonardo DiCaprio   
283                      Drama|Western  Leonardo DiCaprio   
293           Adventure|Drama|Thriller  Leonardo DiCaprio   
294       Biography|Comedy|Crime|Drama  Leonardo DiCaprio   
312                        Crime|Drama  Leonardo DiCaprio   
347               Crime|Drama|Thriller  Leonardo DiCaprio   
431                   Mystery|Thriller  Leonardo DiCaprio   
608              Action|Drama|Thriller  Leonardo DiCaprio   
859              Biography|Crime|Drama  Leonardo DiCaprio   
934           Adventure|Drama|Thriller  Leonardo DiCaprio   
1046                     Drama|Romance  Leonardo DiCaprio   
1336                  Action|Adventure  Leonardo DiCaprio   
1364             Biography|Crime|Drama  Leonardo DiCaprio   
1457           Action|Thriller|Western  Leonardo DiCaprio   
1909                             Drama  Leonardo DiCaprio   
2484                     Drama|Romance  Leonardo DiCaprio   
3025                     Drama|Romance  Leonardo DiCaprio   

                    movie_title  num_voted_users  num_user_for_reviews  \
25                     Titanic            793059                2528.0   
49            The Great Gatsby            362912                 753.0   
94                   Inception           1468200                2803.0   
173               The Revenant            406020                1188.0   
246                The Aviator            264318                 799.0   
283           Django Unchained            955174                1193.0   
293              Blood Diamond            400292                 657.0   
294    The Wolf of Wall Street            780588                1138.0   
312          Gangs of New York            314033                1166.0   
347               The Departed            873649                2054.0   
431             Shutter Island            786092                 964.0   
608               Body of Lies            174248                 263.0   
859        Catch Me If You Can            525801                 667.0   
934                  The Beach            176169                 548.0   
1046        Revolutionary Road            152591                

In [20]:
Brad_Pitt = movies[movies['actor_1_name'] =='Brad Pitt'] # Include all movies in which Brad_Pitt is the lead
Brad_Pitt

director_name  num_critic_for_reviews        gross  \
97            David Fincher                   362.0  127490802.0   
142       Wolfgang Petersen                   220.0  133228348.0   
243       Steven Soderbergh                   198.0  125531634.0   
244              Doug Liman                   233.0  186336103.0   
367              Tony Scott                   142.0      26871.0   
383       Steven Soderbergh                   186.0  183405771.0   
448              David Ayer                   406.0   85707116.0   
579     Jean-Jacques Annaud                    76.0   37901509.0   
646           David Fincher                   315.0   37023395.0   
749         Patrick Gilmore                    98.0   26288320.0   
887             Neil Jordan                   120.0  105264608.0   
1397        Terrence Malick                   584.0   13303319.0   
1608         Andrew Dominik                   273.0    3904982.0   
2030  Alejandro G. Iñárritu                   285.0   34300771.0   
2146    Angelina Jolie Pitt                   131.0     531009.0   
2428         Andrew Dominik                   414.0   14938570.0   
2602             Tony Scott                   122.0   12281500.0   

                                                 genres actor_1_name  \
97                                Drama|Fantasy|Romance    Brad Pitt   
142                                           Adventure    Brad Pitt   
243                                      Crime|Thriller    Brad Pitt   
244                Action|Comedy|Crime|Romance|Thriller    Brad Pitt   
367                               Action|Crime|Thriller    Brad Pitt   
383                                      Crime|Thriller    Brad Pitt   
448                                    Action|Drama|War    Brad Pitt   
579               Adventure|Biography|Drama|History|War    Brad Pitt   
646                                               Drama    Brad Pitt   
749   Adventure|Animation|Comedy|Drama|Family|Fantas...    Brad Pitt   
887                                Drama|Fantasy|Horror    Brad Pitt   
1397                                      Drama|Fantasy    Brad Pitt   
1608              Biography|Crime|Drama|History|Western    Brad Pitt   
2030                                              Drama    Brad Pitt   
2146                                      Drama|Romance    Brad Pitt   
2428                                     Crime|Thriller    Brad Pitt   
2602                Action|Crime|Drama|Romance|Thriller    Brad Pitt   

                                            movie_title  num_voted_users  \
97                 The Curious Case of Benjamin Button            459346   
142                                               Troy            381672   
243                                     Ocean's Twelve            284852   
244                                   Mr. & Mrs. Smith            348861   
367                                           Spy Game            121259   
383                                     Ocean's Eleven            402645   
448                                               Fury            303185   
579                               Seven Years in Tibet             96385   
646                                         Fight Club           1347461   
749                   Sinbad: Legend of the Seven Seas             36144   
887   Interview with the Vampire: The Vampire Chroni...           239752   
1397                                  The Tree of Life            136367   
1608  The Assassination of Jesse James by the Coward...           136104   
2030                                             Babel            243799   
2146                                        By the Sea              7976   
2428                               Killing Them Softly            111625   
2602                                      True Romance            163492   

      num_user_for_reviews language       budget  title_year  imdb_score  \
97                   822.0  English  150000000.

In [21]:
# Write your code for combining the three dataframes here
Combined = (Meryl_Streep.append(Brad_Pitt)).append(Leo_Caprio)
Combined

director_name  num_critic_for_reviews        gross  \
392            Nancy Meyers                   187.0  112703470.0   
1038          Curtis Hanson                    42.0   46815748.0   
1132            Nora Ephron                   252.0   94125426.0   
1322          David Frankel                   208.0  124732962.0   
1390         Robert Redford                   227.0   14998070.0   
1471         Sydney Pollack                    66.0   87100000.0   
1514          David Frankel                   234.0   63536011.0   
1563          Carl Franklin                    64.0   23209440.0   
1784         Stephen Daldry                   174.0   41597830.0   
2500         Phyllida Lloyd                   331.0   29959436.0   
2793          Robert Altman                   211.0   20338609.0   
97            David Fincher                   362.0  127490802.0   
142       Wolfgang Petersen                   220.0  133228348.0   
243       Steven Soderbergh                   198.0  125531634.0   
244              Doug Liman                   233.0  186336103.0   
367              Tony Scott                   142.0      26871.0   
383       Steven Soderbergh                   186.0  183405771.0   
448              David Ayer                   406.0   85707116.0   
579     Jean-Jacques Annaud                    76.0   37901509.0   
646           David Fincher                   315.0   37023395.0   
749         Patrick Gilmore                    98.0   26288320.0   
887             Neil Jordan                   120.0  105264608.0   
1397        Terrence Malick                   584.0   13303319.0   
1608         Andrew Dominik                   273.0    3904982.0   
2030  Alejandro G. Iñárritu                   285.0   34300771.0   
2146    Angelina Jolie Pitt                   131.0     531009.0   
2428         Andrew Dominik                   414.0   14938570.0   
2602             Tony Scott                   122.0   12281500.0   
25            James Cameron                   315.0  658672302.0   
49             Baz Luhrmann                   490.0  144812796.0   
94        Christopher Nolan                   642.0  292568851.0   
173   Alejandro G. Iñárritu                   556.0  183635922.0   
246         Martin Scorsese                   267.0  102608827.0   
283       Quentin Tarantino                   765.0  162804648.0   
293            Edward Zwick                   166.0   57366262.0   
294         Martin Scorsese                   606.0  116866727.0   
312         Martin Scorsese                   233.0   77679638.0   
347         Martin Scorsese                   352.0  132373442.0   
431         Martin Scorsese                   490.0  127968405.0   
608            Ridley Scott                   238.0   39380442.0   
859        Steven Spielberg                   194.0  164435221.0   
934             Danny Boyle                   118.0   39778599.0   
1046             Sam Mendes                   323.0   22877808.0   
1336        Randall Wallace                    83.0   56876365.0   
1364         Clint Eastwood                   392.0   37304950.0   
1457              Sam Raimi                    63.0   18636537.0   
1909             Jerry Zaks                    45.0   12782508.0   
2484           Baz Luhrmann                   106.0   46338728.0   
3025           Baz Luhrmann                   490.0  144812796.0   

                                                 genres       actor_1_name  \
392                                Comedy|Drama|Romance       Meryl Streep   
1038                    Action|Adventure|Crime|Thriller       Meryl Streep   
1132                            Biography|Drama|Romance       Meryl Streep   
1322                               Comedy|Drama|Romance       Meryl Streep   
1390                                 Drama|Thriller|War       Meryl Streep   
1471                            Biography|Drama|Romance       Meryl Streep   
1514                               Comedy|Drama|Romance       Meryl

In [22]:
# Write your code for grouping the combined dataframe here
num_critic_for_reviews = Combined.groupby(["actor_1_name"])['num_critic_for_reviews'].mean()
print(num_critic_for_reviews)
num_user_for_reviews = Combined.groupby(["actor_1_name"])['num_user_for_reviews'].mean()
print(num_user_for_reviews)

actor_1_name
Brad Pitt            245.000000
Leonardo DiCaprio    330.190476
Meryl Streep         181.454545
Name: num_critic_for_reviews, dtype: float64
actor_1_name
Brad Pitt            742.352941
Leonardo DiCaprio    914.476190
Meryl Streep         297.181818
Name: num_user_for_reviews, dtype: float64


#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [25]:
print(num_user_for_reviews)


actor_1_name
Brad Pitt            742.352941
Leonardo DiCaprio    914.476190
Meryl Streep         297.181818
Name: num_user_for_reviews, dtype: float64


#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [24]:
print(num_critic_for_reviews)

actor_1_name
Brad Pitt            245.000000
Leonardo DiCaprio    330.190476
Meryl Streep         181.454545
Name: num_critic_for_reviews, dtype: float64
